### PR TITLE
fix(upstreams) deduplicate like targets

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:

Combine targets with the same address or hostname by summing their weights into a single target. Previously, same address targets would result in duplicates and would be rejected by Kong.

**Which issue this PR fixes**:

Fix https://github.com/Kong/kubernetes-ingress-controller/issues/5761. Summing weights wasn't actually that hard!

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/translator/translator.go
Comment (pmalek) [inline internal/dataplane/translator/translator.go:425]: `targetMap` maps from ? (`string`)  to ... IPs ? I'd be inclined to say that naming ( I know ) might be worth improving here.

Perhaps put `*.t.Target.Target` in a named variable? `ip`? `targetIP ?`

File: internal/dataplane/translator/translator.go
Comment (pmalek) [inline internal/dataplane/translator/translator.go:427]: Apart from naming there's `existing.Weight` and `existing.Target.Weight` which is not immediately clear what's what here.

Do you mind clueing the readers on what's happening here?

File: internal/dataplane/translator/translator.go
Comment (pmalek) [inline internal/dataplane/translator/translator.go:1]: A handful of tests would definitely be helpful to show show that the change works as intended.

File: internal/dataplane/translator/translator.go
Comment (rainest) [reply internal/dataplane/translator/translator.go:425]: Blame Kong upstream. The latest generated docs are unfortunately still kind of a mess, but it's literally just "target": https://docs.konghq.com/gateway/3.4.x/admin-api/#target-object

`target` is a hostname _or_ IP and its port, delimited by a colon, e.g. `example.com:80` or `10.0.01:9000` (I assume IPv6 uses `[fe80::1]:80` but didn't confirm). Ultimately it's _some_ unique string, and we trust that the endpoint resolver will give us a string that's a valid Kong target. We don't care about the details as long as it's unique. Will add a bit more to the comment.

File: internal/dataplane/translator/translator.go
Comment (rainest) [reply internal/dataplane/translator/translator.go:1]: We did not previously have a unit test for `getUpstreams()` and, after attempting one, I can see why: it relies a lot on previous translator steps filling in all the various inputs, and trying to build all those by hand in a test case is quite painful. [test_getupstreams.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/14910994/test_getupstreams.diff.txt) has the attempt before I gave up trying to make it not segfault for lack of prior steps populating everything.

Integration would be hard here since we either have to look at config (AFAIK we don't do this in integration) or somehow infer from resolved upstream addresses that we've assigned the correct weights, which is a pretty poor test of what this is doing.

In light of that, I broke the deduplication code out into a helper, since that makes the unit test much simpler.

**Comment by rainest** [comment]
UAT confirmed this appears to resolve the issue. Added unit tests, should be good to go for review now.

File: internal/dataplane/translator/translator.go
Comment (czeslavo) [inline internal/dataplane/translator/translator.go:1]: Given we're already touching this code, I think it could be worth extracting `Upstream`-related helpers to a separate file (e.g., `translator_upstreams.go`) to leave the `translator.go` for the top-level methods only. This can be done in a separate PR to not block this one.

**Review by czeslavo — APPROVED** [review]

**Comment by mithunkrb** [comment]
Hey @rainest , Thanks a lot for fixing this issue. We are also encountering this issue when using KIC with argo rollouts(using canary), wondering when can we expect the next release/patch release of kic having this fix?

**Comment by rainest** [comment]
@mithunkrb probably in 3.2, with a rough timeline of ~1.5months out.

https://hub.docker.com/r/kong/nightly-ingress-controller/tags?page=1&ordering=last_updated is available if you want it earlier.

Note that we did find [a bug](https://github.com/Kong/kubernetes-ingress-controller/pull/5946) in the fix, so you may want to wait til the build after that merges. Not sure what all it affects, but it definitely affects any Services that use the `ingress.kubernetes.io/service-upstream` annotation.

**Comment by mithunkrb** [comment]
Thank you @rainest  for the update, it helps! For the time being would go with the nightly build(once its available) having the fix for the mentioned [bug](https://github.com/Kong/kubernetes-ingress-controller/pull/5946), and would upgrade to 3.2 once we have that available.

**Pull request merged**
